### PR TITLE
Correctly handle converters and default values

### PR DESCRIPTION
**Pull request by kevmoo**
A consistent model for handling converter-like things and default values.

Also simplified and DRY'd up a number of code paths.

### CODE REVIEW

File: json_serializable/lib/src/decode_helper.dart [hunk 1]
```diff
@@ -227,16 +227,6 @@ abstract class DecodeHelper implements HelperCore {
             'effect because both `disallowNullValue` and `required` are set to '
             '`true`.');
       }
-
-      final deserializeConvertData = contextHelper.deserializeConvertData;
-
-      if (deserializeConvertData != null &&
-          !deserializeConvertData.nullableToAllowDefault) {
-        log.warning('The field `${field.name}` has both `defaultValue` and '
```
Comment (kevmoo) [inline json_serializable/lib/src/decode_helper.dart:235]: We now support this case! Nice!

File: json_serializable/lib/src/default_container.dart [hunk 2]
```diff
@@ -27,6 +29,10 @@ class DefaultContainer {
       return input._result(defaultValue);
     }
 
+    if (input is LambdaResult && defaultValue != null) {
```
Comment (kevmoo) [inline json_serializable/lib/src/default_container.dart:32]: `LambdaResult` is used by the converters – so we can catch it here and handle the default case!

File: json_serializable/lib/src/lambda_result.dart [hunk 3]
```diff
@@ -2,17 +2,23 @@
...
 // BSD-style license that can be found in the LICENSE file.
 
+import 'constants.dart';
+
 class LambdaResult {
   final String expression;
   final String lambda;
+  final String? asContent;
+
+  String get _fullExpression => '$expression${asContent ?? ''}';
```
Comment (kevmoo) [inline json_serializable/lib/src/lambda_result.dart:12]: Need `expression` to remain "pure" so we can do the `null` check.

File: json_serializable/lib/src/type_helper_ctx.dart [hunk 4]
```diff
@@ -125,8 +125,6 @@ ConvertData? _convertData(DartObject obj, FieldElement element, bool isFrom) {
         'positional parameter.');
   }
 
-  var nullableToAllowDefault = false;
```
Comment (kevmoo) [inline json_serializable/lib/src/type_helper_ctx.dart:128]: no longer needed!

File: json_serializable/lib/src/json_key_utils.dart [hunk 5]
```diff
@@ -202,7 +202,7 @@ KeyConfig _from(FieldElement element, ClassConfig classAnnotation) {
 KeyConfig _populateJsonKey(
   ClassConfig classAnnotation,
   FieldElement element, {
-  required Object? defaultValue,
+  required String? defaultValue,
```
Comment (kevmoo) [inline json_serializable/lib/src/json_key_utils.dart:205]: Turns out we can be more specific here!

File: json_serializable/lib/src/type_helper.dart [hunk 6]
```diff
@@ -91,5 +92,5 @@ Object commonNullPrefix(
   Object unsafeExpression,
 ) =>
     nullable
-        ? '$expression == null ? null : $unsafeExpression'
+        ? ifNullOrElse(expression, 'null', '$unsafeExpression')
```
Comment (kevmoo) [inline json_serializable/lib/src/type_helper.dart:95]: Have this pattern three times, make it a shared util.

**Review by natebosch — APPROVED** [review]

**Pull request merged**